### PR TITLE
tier: Add force flag to remove a tier

### PR DESCRIPTION
**Pull request by vadmeste**
Sometimes, the tier is down or its address change and we could have 
many reasons why it is not possible to reach the tier server during removal 
of a tier. For that, we need to add a force flag.

### CODE REVIEW

**Comment by harshavardhana** [comment]
what are the consequences of doing this wrong?

**Comment by vadmeste** [comment]
> what are the consequences of doing this wrong?

for now, it is not possible to re-add the tier again. Currently if the tier contains some data, then the removal will not work until the user removes the data manually from the tier. However, there are some other corner cases, like the tier is down (like a user decided to kill the tiering itself). You can never remove this tier configuration in MinIO side in that case.

**Comment by harshavardhana** [comment]
> for now, it is not possible to re-add the tier again. Currently if the tier contains some data, then the removal will not work until the user removes the data manually from the tier. However, there are some other corner cases, like the tier is down (like a user decided to kill the tiering itself). You can never remove this tier configuration in MinIO side in that case.

yes but this is disastrous, all the data that exists on the hot-tier will become stale and dangling. 

I don't think we need this feature at all.

**Comment by vadmeste** [comment]
> I don't think we need this feature at all.

Well, it is a force flag, we can require --dangerous in addition to --force in mc. 

`mc ilm tier remove <ALIAS> AZTIER --force --dangerous`

I said there are some scenarios where you cannot remove the existing tier anymore, like a user who shutdowns the tier before removing the tier config in MinIO, the tier configuration will be stuck in MinIO forever.

**Comment by harshavardhana** [comment]
> > I don't think we need this feature at all.
> 
> Well, it is a force flag, we can require --dangerous in addition to --force in mc.
> 
> `mc ilm tier remove <ALIAS> AZTIER --force --dangerous`
> 
> I said there are some scenarios where you cannot remove the existing tier anymore, like a user who shutdowns the tier before removing the tier config in MinIO, the tier configuration will be stuck in MinIO forever.

Then we can make this a hidden feature in mc, there is no good reason user should do that.

It's a data loss scenario and it's not a daily use behavior.

File: tier.go [hunk 1]
```diff
@@ -155,12 +155,17 @@ func (adm *AdminClient) EditTier(ctx context.Context, tierName string, creds Tie
 }
 
 // RemoveTier removes an empty tier identified by tierName
-func (adm *AdminClient) RemoveTier(ctx context.Context, tierName string) error {
+func (adm *AdminClient) RemoveTier(ctx context.Context, tierName string, force bool) error {
```
Comment (harshavardhana) [inline tier.go:158]: Please don't break existing APIs write new ones.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: tier.go [hunk 2]
```diff
@@ -177,6 +177,35 @@ func (adm *AdminClient) RemoveTier(ctx context.Context, tierName string) error {
 	return nil
 }
 
+// RemoveTierV2 removes an empty tier identified by tierName, the tier is not
+// required to be reachable or empty  if force flag is set to true.
+func (adm *AdminClient) RemoveTierV2(ctx context.Context, tierName string, force bool) error {
```
Comment (harshavardhana) [inline tier.go:182]:
```suggestion
func (adm *AdminClient) RemoveTierV2(ctx context.Context, opts TierOptions) error {
```

This is better for future extensibility.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: tier.go (hunk 2 above)
Comment (vadmeste) [reply tier.go]: what about now ? (little different to what you suggested)

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**